[FLINK-17844][build] Enforce @PublicEvolving compatibility for minor versions

zentol opened the pull request:
This PR makes 2 changes to the `create_release_branch` release script:

1) The japicmp reference version is now automatically updated.
   - For a major release x.y.0 the reference version is x.y-1.0
      - 1.10.0 checks against 1.9.0
   - For a minor release x.y.z the reference version is x.y.z-1
      - 1.10.1 checks against 1.10.0

2) Enforce `@PublicEvolving` compatibility for minor versions
   - remove japicmp exclusion
   - un-comment the newly added japicmp inclusion

rmetzger commented on tools/releasing/create_release_branch.sh:
```diff
@@ -59,6 +59,22 @@ git checkout -b $target_branch
+  # harden compatibility constraints for minor versions
+  perl -pi -e 's#<!--(<include>\@org.apache.flink.annotation.PublicEvolving</include>)-->#${1}#' pom.xml
+  perl -pi -e 's#\t+<exclude>\@org.apache.flink.annotation.PublicEvolving.*\n##' pom.xml
+fi
```
Shouldn't these changes go into `releasing/update_branch_version.sh`? This file here seems to be used for preparing release candidate branches. (According to https://cwiki.apache.org/confluence/display/FLINK/Creating+a+Flink+Release#CreatingaFlinkRelease-Createareleasebranch)

zentol replied:
hmm, you're right, but I think we should have both.

For the release-X.Y branches we have to update the configuration manually, so this PR is more of a final safeguard for actual releases in case the the update is forgotten, which is basically guaranteed to happen at some point.

rmetzger replied:
Ah, so the idea is that the build of a `release-x.y.z-rci` branch would fail (as a last resort) because the japicmp plugin gets reconfigured as part of the release process.
But ideally the release managers manually uncomment this one line in the pom in the `release-X.Y` branch.

zentol replied:
It wasn't the original idea, I just remembered wrong how the release process works.

As for the manual work, the RM has to un-comment the line, remove the exclusion and update the reference version to the last release minor version.

Ultimately the logic is identical, so I'll see if I can refactor this update into a separate script that the RM can call.

rmetzger commented on pom.xml:
```diff
@@ -1961,6 +1961,8 @@ under the License.
 							<onlyModified>true</onlyModified>
 							<includes>
 								<include>@org.apache.flink.annotation.Public</include>
+								<!-- The following line must be un-commented for minor releases-->
```
Maybe update the comment that this will be done automatically by a script?

zentol commented on tools/releasing/create_release_branch.sh:
```diff
@@ -59,6 +59,22 @@ git checkout -b $target_branch
 
+ultra=$(echo "${NEW_VERSION}" | perl -p -e 's#(\d+)\.\d+\.\d+#$1#')
+major=$(echo "${NEW_VERSION}" | perl -p -e 's#\d+\.(\d+)\.\d+#$1#')
+minor=$(echo "${NEW_VERSION}" | perl -p -e 's#\d+\.\d+\.(\d+)#$1#')
```
no;  we use ?.major.minor unfortunately.

rmetzger replied:
I know that we are not using the right terminologies in Flink, but maybe this is our chance of slowly introducing new terms.

zentol replied:
Doing that will just introduce confusion; recent discussions on the ML showed that several devs were not aware of our version scheme, so it could easily happen that another one sees this and thinks that's what we are using.

rmetzger replied:
+1

zentol commented:
A separate profile is imo overkill, when all we really need is to move one entry.

rmetzger commented:
I'm not saying you should do it the way I would have done it :) And yeah, it would be a lot of noise in our pom files for such a tiny change.

zentol commented:
@rmetzger I've updated the PR.

I've added `tools\releasing\update_japicmp_configuration.sh` to handle the configuration updates for both releases and master/snapshot branches. The file also contains a description for which scenarios it should be used for.
It covers all cases without any extra configuration from the outside; so the RM just has to call it and commit the changes.

rmetzger commented on tools/releasing/update_japicmp_configuration.sh:
```diff
@@ -0,0 +1,83 @@
+    # this is a safeguard in case the manual step of enabling checks after the X.Y.0 release was forgotten
+    enable_public_evolving_compatibility_checks
+  fi
+elif [[ ${current_branch} =~ -SNAPSHOT ]]; then
```
when do we have a branch name containing "-SNAPSHOT"? Shouldn't this be a check on the version set in the ./pom.xml ?

zentol replied:
When for example we just released 1.11.0 and want to updated 1.11-SNAPSHOT to check against 1.11.0 and enable compatibility guarantees .

rmetzger replied:
Will also be a manual invocation of the script described in the release guide?
The RM will have to rename `release-1.11` to `release-1.11-SNAPSHOT` and then run this script after 1.11.0 has been released?

I still wonder why this check is on the branch name, and not the version.

zentol replied:
ah right, the branch name does not contain snapshot doesn't it. hmm :/

We do need to take the branch name into consideration since we never want to enable stronger guarantees on the master branch.

zentol replied:
Now checks for a `release-` prefix, after the rc/master checks.

rmetzger commented on tools/releasing/update_japicmp_configuration.sh:
```diff
@@ -0,0 +1,83 @@
+  # snapshot branch
+  set_japicmp_reference_version ${NEW_VERSION}
+  enable_public_evolving_compatibility_checks
+elif [[ ${current_branch} =~ ^master$ ]]; then
```
This case (and the previous one) require a manual invocation of `update_japicmp_configuration.sh` by the RM, right? (I assume you'll update the wiki?)

zentol replied:
correct

rmetzger commented on tools/releasing/update_japicmp_configuration.sh:
```diff
@@ -0,0 +1,83 @@
+
+current_branch=$(git rev-parse --abbrev-ref HEAD)
+
+if [[ ${current_branch} =~ -rc ]]; then
```
for this case, the RM doesn't need to invoke this script, because it is called from the `create_release_branch.sh` script, right?

zentol replied:
correct
